File: src/CycleSort.java
public class CycleSort {
    // Main function to test the Cycle sort function
    public static void main(String[] args) {
        int arr[] = {1, 8, 3, 9, 10, 10, 2, 4};
        int n = arr.length;
        cycleSort(arr, n);

        System.out.println("After sort: ");
        for (int i = 0; i < n; i++) {
            System.out.println(arr[i] + " ");
        }
    }

    // Function sort the array using Cycle sort
    public static void cycleSort(int arr[], int n) {
        int w = 0; // count number of memory writes

        for (int start = 0; start <= n - 2; start++) {
            int item = arr[start]; // initializing the starting point

            int pos = start; // find the position where we put the item
            for (int i = start + 1; i < n; i++)
                if (arr[i] < item)
                    pos++;

            if (pos == start) // if the item is already in the correct position
                continue;

            while (item == arr[pos]) // ignore all duplicate elements
                pos += 1;

            if (pos != start) { // put the item to it's right position
                int temp = item;
                item = arr[pos];
                arr[pos] = temp;
                w++;
            }

            while (pos != start) { // rotate rest of the cycle
                pos = start;

                for (int i = start + 1; i < n; i++)
                    if (arr[i] < item)
                        pos += 1;

                while (item == arr[pos])
                    pos += 1;

                if (item != arr[pos]) {
                    int temp = item;
                    item = arr[pos];
                    arr[pos] = temp;
                    w++;
                }
            }
        }
    }
}

File: Cpp/Cyclesort.cpp
#include <iostream>
#include <vector>

void cycleSort(std::vector<int> &arr) {
    for (int currentIndex = 0; currentIndex < arr.size() - 1; currentIndex++) {
        int item = arr[currentIndex];
        int currentIndexCopy = currentIndex;

        for (int i = currentIndex + 1; i < arr.size(); i++)
            if (arr[i] < item)
                currentIndexCopy++;
      
        if (currentIndexCopy == currentIndex)
            continue;

        while (item == arr[currentIndexCopy])
            currentIndexCopy++;

        int temp = arr[currentIndexCopy];
        arr[currentIndexCopy] = item;
        item = temp;

        while (currentIndexCopy != currentIndex) {

            currentIndexCopy = currentIndex;

            for (int i = currentIndex + 1; i < arr.size(); i++)
                if (arr[i] < item)
                    currentIndexCopy++;

            while (item == arr[currentIndexCopy])
                currentIndexCopy++;

            temp = arr[currentIndexCopy];
            arr[currentIndexCopy] = item;
            item = temp;
        }
    }
}
